sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,1582 @@
+# because we need to provide the dtype of the vector but can't use numeric fused types.
+cdef class StdVectorSentinelDTYPE(StdVectorSentinel):
+    cdef vector[DTYPE_t] vec
+
+    @staticmethod
+    cdef StdVectorSentinel create_for(vector[DTYPE_t] * vec_ptr):
+        # This initializes the object directly without calling __init__
+        cdef StdVectorSentinelDTYPE sentinel = StdVectorSentinelDTYPE.__new__(StdVectorSentinelDTYPE)
+        sentinel.vec.swap(deref(vec_ptr))
+        return sentinel
+
+
+cdef class StdVectorSentinelITYPE(StdVectorSentinel):
+    cdef vector[ITYPE_t] vec
+
+    @staticmethod
+    cdef StdVectorSentinel create_for(vector[ITYPE_t] * vec_ptr):
+        # This initializes the object directly without calling __init__
+        cdef StdVectorSentinelITYPE sentinel = StdVectorSentinelITYPE.__new__(StdVectorSentinelITYPE)
+        sentinel.vec.swap(deref(vec_ptr))
+        return sentinel
+
+
+cdef np.ndarray vector_to_nd_array(vector_DITYPE_t * vect_ptr):
+    """Create a numpy ndarray given a C++ vector.
+
+    The numpy array buffer is the one of the C++ vector.
+    A StdVectorSentinel is registered as the base object for the numpy array,
+    freeing the C++ vector it encapsulates when the numpy array is freed.
+    """
+    typenum = DTYPECODE if vector_DITYPE_t is vector[DTYPE_t] else ITYPECODE
+    cdef:
+        np.npy_intp size = deref(vect_ptr).size()
+        # TODO: use PyArray_SimpleNewFromData from the Numpy C API directly
+        # I've tried, but Cython fails when parsing the C API
+        np.ndarray arr = np.PyArray_SimpleNewFromData(1, &size, typenum,
+                                                      deref(vect_ptr).data())
+        StdVectorSentinel sentinel
+
+    if vector_DITYPE_t is vector[DTYPE_t]:
+        sentinel = StdVectorSentinelDTYPE.create_for(vect_ptr)
+    else:
+        sentinel = StdVectorSentinelITYPE.create_for(vect_ptr)
+
+    # Makes the numpy array responsible of the life-cycle of its buffer.
+    # A reference to the StdVectorSentinel will be stolen by the call bellow,
+    # so we increase its reference counter.
+    # See: https://docs.python.org/3/c-api/intro.html#reference-count-details
+    Py_INCREF(sentinel)
+    PyArray_SetBaseObject(arr, <PyObject*>sentinel)
+    return arr
+
+
+cdef np.ndarray[object, ndim=1] coerce_vectors_to_nd_arrays(
+    vector_vector_DITYPE_t* vecs
+):
+    """Coerce a std::vector of std::vector to a ndarray of ndarray."""
+    cdef:
+        ITYPE_t n = deref(vecs).size()
+        np.ndarray[object, ndim=1] nd_arrays_of_nd_arrays = np.empty(n,
+                                                                     dtype=np.ndarray)
+
+    for i in range(n):
+        nd_arrays_of_nd_arrays[i] = vector_to_nd_array(&(deref(vecs)[i]))
+
+    return nd_arrays_of_nd_arrays
+
+#####################
+
+cdef class PairwiseDistancesReduction:
+    f"""Abstract class which compute pairwise distances between
```

Comment:
I'd advise not to use fstring for docstrings since it makes the ``__doc__`` attribute empty. For a simple use case like here, I'd just put the value of the default chunk size directly